scripts/sentiment_analysis/finetune_lm.py

```diff
@@ -193,152 +241,130 @@ def preprocess_dataset(dataset):
-            print('Observed Improvement.')
-            stop_early = 0
-            net.save_parameters(args.save_prefix + '_{:04d}.params'.format(epoch))
-            best_valid_acc = valid_acc
-
-    net.load_parameters(glob.glob(args.save_prefix+'_*.params')[-1], context)
-    valid_avg_L, valid_acc = evaluate(valid_dataloader)
-    test_avg_L, test_acc = evaluate(test_dataloader)
-    print('Best validation loss %g, validation acc %.4f'%(valid_avg_L, valid_acc))
-    print('Best test loss %g, test acc %.4f'%(test_avg_L, test_acc))
-    print('Total time cost %.2fs'%(time.time()-start_pipeline_time))
-
+trainer = gluon.Trainer(net.collect_params(), 'ftml',
+                        {'learning_rate': args.lr})
+loss_fn = gluon.loss.SigmoidBCELoss()
+
+
+class SentimentAnalysisBatchProcessor(BatchProcessor):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def evaluate_batch(self, estimator, val_batch, batch_axis=0):
+        data = mx.nd.transpose(val_batch[0][0])
+        valid_length = val_batch[0][1].astype(np.float32)
+        label = val_batch[1]
+        data_list = gluon.utils.split_and_load(data, estimator.context)
+        valid_length_list = gluon.utils.split_and_load(valid_length,
+                                                       estimator.context)
+        label_list = gluon.utils.split_and_load(label, estimator.context)
+        pred = [
+            estimator.eval_net(data, valid_length)
+            for data, valid_length in zip(data_list, valid_length_list)
+        ]
+        loss = [
+            estimator.evaluation_loss(y_hat, y)
+            for y_hat, y in zip(pred, label_list)
+        ]
+        pre_labels = [(pred_pro > 0.5).reshape(-1, ) for pred_pro in pred]
+        return data_list, label_list, pre_labels, loss
+
+    def fit_batch(self, estimator, train_batch, batch_axis=0):
+        data = mx.nd.transpose(train_batch[0][0])
+        valid_length = train_batch[0][1].astype(np.float32)
+        label = train_batch[1]
+        data_list = gluon.utils.split_and_load(data, estimator.context)
+        valid_length_list = gluon.utils.split_and_load(valid_length,
+                                                       estimator.context)
+        label_list = gluon.utils.split_and_load(label, estimator.context)
+        with autograd.record():
+            pred = [
+                estimator.net(data, valid_length)
+                for data, valid_length in zip(data_list, valid_length_list)
+            ]
+            loss = [
+                estimator.loss(y_hat, y).mean()
+                for y_hat, y in zip(pred, label_list)
+            ]
+        for l in loss:
+            l.backward()
+        # Clip gradient
+        if args.clip is not None:
+            grads = [
+                p.grad(estimator.context)
+                for p in self.net.collect_params().values()
+            ]
+            gluon.utils.clip_global_norm(grads, args.clip)
```

Comment:
Because you split data and labels across multiple contexts, you probably need to call `allreduce_grads()` after gradient clipping before back propogation.